internal/telemetry/telemetry.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2022 MongoDB Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telemetry
+
+import (
+	"encoding/json"
+	"errors"
+	"os"
+	"path"
+
+	"github.com/mongodb/mongocli/internal/config"
+	"github.com/spf13/afero"
+)
+
+const (
+	dirPermissions   = 0700
+	filePermissions  = 0600
+	cacheFilename    = "telemetry"
+	maxCacheFileSize = 100 * 1024 * 1024 // 100MB
+)
+
+var fs = afero.NewOsFs()
+
+type Event struct {
+	Timestamp  string            `json:"timestamp"`
+	Source     string            `json:"source"`
+	Name       string            `json:"name"`
+	Properties map[string]string `json:"properties"`
+}
+
+func (event *Event) Cache() error {
+	if !config.TelemetryEnabled() {
+		return nil
+	}
+	file, err := openCacheFile()
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	data, _ := json.MarshalIndent(event, "", "\t")
+	_, err = file.Write(data)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func openCacheFile() (afero.File, error) {
+	cacheDir, err := os.UserCacheDir()
```

Comment:
the main point of using afero is for us to mock file system interactions during unit tests, righ now this implementation uses both direct file system (via `os`) and afero interactions, see this comment https://github.com/mongodb/mongodb-atlas-cli/pull/1134#discussion_r854153786 that PR has a good discussion and resolution to using afero for testing while not exposing it to consumers of the package

Reply:
just as a side note afero has built-in support for cache in fs https://github.com/spf13/afero/blob/100c9a6d7b9b70ddb10dc728b9b22d472e1f7f71/cacheOnReadFs.go see if it makes sense in here